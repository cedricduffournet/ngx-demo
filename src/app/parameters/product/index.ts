export * from './parameter-product.component';
