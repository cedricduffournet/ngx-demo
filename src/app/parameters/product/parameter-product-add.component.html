<app-product-add-view></app-product-add-view>
