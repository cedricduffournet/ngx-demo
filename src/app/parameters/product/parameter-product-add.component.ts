import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-parameter-product-add',
  templateUrl: './parameter-product-add.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ParameterProductAddComponent implements OnInit {
  constructor() {}

  ngOnInit() {}
}
