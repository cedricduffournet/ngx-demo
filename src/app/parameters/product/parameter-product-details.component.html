<app-product-details-view></app-product-details-view>
