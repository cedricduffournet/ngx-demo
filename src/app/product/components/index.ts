export * from './product-add/product-add.component';
export * from './product-update/product-update.component';
export * from './product-delete/product-delete.component';
export * from './product-form/product-form.component';
export * from './product-item/product-item.component';
export * from './product-items/product-items.component';
export * from './product-details/product-details.component';
