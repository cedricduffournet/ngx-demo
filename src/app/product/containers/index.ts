export * from './product-add-view/product-add-view.component';
export * from './product-update-modal/product-update-modal.component';
export * from './product-delete-modal/product-delete-modal.component';
export * from './product-list-view/product-list-view.component';
export * from './product-selected/product-selected.component';
export * from './product-details-view/product-details-view.component';
