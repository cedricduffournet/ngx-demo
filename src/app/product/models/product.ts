import { schema } from 'normalizr';

export interface Product {
  id: number;
  name: string;
}

export const productSchema = new schema.Entity('products');
