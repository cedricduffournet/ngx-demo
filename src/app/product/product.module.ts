import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { ProductEffects } from '@app/product/state/effects';
import { ProductFacade } from '@app/product/state/product.facade';
import {
  ProductAddComponent,
  ProductUpdateComponent,
  ProductDeleteComponent,
  ProductFormComponent,
  ProductItemComponent,
  ProductItemsComponent
} from '@app/product/components';
import {
  ProductAddModalComponent,
  ProductUpdateModalComponent,
  ProductDeleteModalComponent,
  ProductListViewComponent
} from '@app/product/containers';
import { reducers } from '@app/product/state/reducers';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    EffectsModule.forFeature([ProductEffects]),
    StoreModule.forFeature('products', reducers)
  ],
  declarations: [
    ProductListViewComponent,
    ProductAddComponent,
    ProductUpdateComponent,
    ProductDeleteComponent,
    ProductFormComponent,
    ProductItemComponent,
    ProductItemsComponent,
    ProductAddModalComponent,
    ProductUpdateModalComponent,
    ProductDeleteModalComponent
  ],
  entryComponents: [
    ProductUpdateModalComponent,
    ProductAddModalComponent,
    ProductDeleteModalComponent
  ],
  exports: [ProductListViewComponent],
  providers: [ProductFacade]
})
export class ProductModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
