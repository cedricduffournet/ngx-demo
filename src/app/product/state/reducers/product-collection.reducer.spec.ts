import { reducer } from '@app/product/state/reducers/product-collection.reducer';
import * as fromProducts from '@app/product/state/reducers/product-collection.reducer';

import {
  ProductListViewActions,
  ProductAddModalActions,
  ProductDeleteModalActions,
  ProductApiActions
} from '@app/product/state/actions';
import { Product } from '@app/product/models/product';

describe('ProductCollectionReducer', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD', () => {
    it('should set loading to true', () => {
      const action = ProductListViewActions.loadProducts();
      const result = reducer(fromProducts.INITIAL_STATE, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD_SUCCESS ', () => {
    it('should set loading to false, loaded to true and set product ids [1, 2]', () => {
      const initialState = {
        ...fromProducts.INITIAL_STATE,
        loading: true
      };

      const products = {
        entities: {
          1: {
            name: 'Name 1'
          },
          2: {
            name: 'Name 2'
          }
        },
        result: [1, 2]
      };
      const action = ProductApiActions.loadProductSuccess({
        products
      });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD_FAILURE ', () => {
    const initialState = {
      ...fromProducts.INITIAL_STATE,
      loading: true
    };

    it('should set loading to false, loaded to false', () => {
      const error = {
        message: 'error'
      };
      const action = ProductApiActions.loadProductFailure({ error });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('ADD', () => {
    it('should set adding to true on ProductAddModalActions.addProduct', () => {
      const product = {
        name: 'AddName'
      } as Product;
      const action = ProductAddModalActions.addProduct({
        product
      });
      const result = reducer(fromProducts.INITIAL_STATE, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('ADD_SUCCESS', () => {
    it('should set adding to false, added to true and add new product (3) to ids', () => {
      const initialState = {
        ...fromProducts.INITIAL_STATE,
        adding: true,
        ids: [1, 2]
      };
      const product = {
        entities: {
          3: {
            name: 'AddName'
          }
        },
        result: 3
      };
      const action = ProductApiActions.addProductSuccess({
        product
      });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('ADD_FAILURE & SHOW_MODAL_ADD', () => {
    const initialState = {
      ...fromProducts.INITIAL_STATE,
      adding: true,
      added: true
    };

    it('should set adding to false, added to false', () => {
      const error = {
        message: 'error'
      };
      const action = ProductApiActions.addProductFailure({ error });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });

    it('should set adding to false on ProductListViewActions.showAddProductModal', () => {
      const action = ProductListViewActions.showAddProductModal();
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('DELETE', () => {
    it('should set deleting to true', () => {
      const product = {
        id: 1,
        name: 'RemoveName'
      };
      const action = ProductDeleteModalActions.deleteProduct({
        product
      });
      const result = reducer(fromProducts.INITIAL_STATE, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('DELETE_SUCCESS', () => {
    const initialState = {
      ...fromProducts.INITIAL_STATE,
      removing: true,
      ids: [1, 2]
    };

    it('should set deleting to false, deleted to true and remove civproductility (1) form ids', () => {
      const id = 1;

      const action = ProductApiActions.deleteProductSuccess({
        id
      });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('DELETE_FAILURE & SHOW_MODAL_DELETE', () => {
    const initialState = {
      ...fromProducts.INITIAL_STATE,
      removing: true,
      removed: true
    };

    it('should set removing to false, remove to false', () => {
      const error = {
        message: 'error'
      };
      const action = ProductApiActions.deleteProductFailure({ error });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });

    it('should set removing to false, remove to false on ProductListViewActions.showDeleteProductModal', () => {
      const action = ProductListViewActions.showDeleteProductModal();
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('getIds', () => {
      it('should retrieve ids [1,2,3]', () => {
        const result = fromProducts.getIds({
          ...fromProducts.INITIAL_STATE,
          ids: [1, 2, 3]
        });
        expect(result).toMatchSnapshot();
      });
    });

    describe('getDeleting', () => {
      it('should retrieve removing false', () => {
        const result = fromProducts.getDeleting({
          ...fromProducts.INITIAL_STATE,
          deleting: false
        });

        expect(result).toMatchSnapshot();
      });

      it('should retrieve removing true', () => {
        const result = fromProducts.getDeleting({
          ...fromProducts.INITIAL_STATE,
          deleting: true
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getDeleted', () => {
      it('should retrieve deleted false', () => {
        const result = fromProducts.getDeleted({
          ...fromProducts.INITIAL_STATE,
          deleted: false
        });

        expect(result).toMatchSnapshot();
      });

      it('should retrieve deleted true', () => {
        const result = fromProducts.getDeleted({
          ...fromProducts.INITIAL_STATE,
          deleted: true
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getAdding', () => {
      it('should retrieve adding false', () => {
        const result = fromProducts.getAdding({
          ...fromProducts.INITIAL_STATE,
          adding: false
        });

        expect(result).toMatchSnapshot();
      });

      it('should retrieve adding true', () => {
        const result = fromProducts.getAdding({
          ...fromProducts.INITIAL_STATE,
          adding: true
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getAdded', () => {
      it('should retrieve added false', () => {
        const result = fromProducts.getAdded({
          ...fromProducts.INITIAL_STATE,
          added: false
        });

        expect(result).toMatchSnapshot();
      });

      it('should retrieve added true', () => {
        const result = fromProducts.getAdded({
          ...fromProducts.INITIAL_STATE,
          added: true
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getLoaded', () => {
      it('should retrieve loaded false', () => {
        const result = fromProducts.getLoaded({
          ...fromProducts.INITIAL_STATE,
          loaded: false
        });

        expect(result).toMatchSnapshot();
      });

      it('should retrieve loaded true', () => {
        const result = fromProducts.getLoaded({
          ...fromProducts.INITIAL_STATE,
          loaded: true
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getLoading', () => {
      it('should retrieve loading false', () => {
        const result = fromProducts.getLoading({
          ...fromProducts.INITIAL_STATE,
          loading: false
        });

        expect(result).toMatchSnapshot();
      });

      it('should retrieve loading true', () => {
        const result = fromProducts.getLoading({
          ...fromProducts.INITIAL_STATE,
          loading: true
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
