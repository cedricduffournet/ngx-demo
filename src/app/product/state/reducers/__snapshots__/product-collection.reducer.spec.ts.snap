// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCollectionReducer ADD should set adding to true on ProductAddModalActions.addProduct 1`] = `
Object {
  "added": false,
  "adding": true,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer ADD_FAILURE & SHOW_MODAL_ADD should set adding to false on ProductListViewActions.showAddProductModal 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer ADD_FAILURE & SHOW_MODAL_ADD should set adding to false, added to false 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer ADD_SUCCESS should set adding to false, added to true and add new product (3) to ids 1`] = `
Object {
  "added": true,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [
    1,
    2,
    3,
  ],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer CHANGE_PAGE current page shoud be set to 10 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 10,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer DELETE should set deleting to true 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": true,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer DELETE_FAILURE & SHOW_MODAL_DELETE should set removing to false, remove to false 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "removed": true,
  "removing": true,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer DELETE_FAILURE & SHOW_MODAL_DELETE should set removing to false, remove to false on ProductListViewActions.showDeleteProductModal 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "removed": true,
  "removing": true,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer DELETE_SUCCESS should set deleting to false, deleted to true and remove civproductility (1) form ids 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": true,
  "deleting": false,
  "ids": Array [
    2,
  ],
  "loaded": false,
  "loading": false,
  "removing": true,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer LOAD should set loading to true 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": true,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer LOAD_FAILURE  should set loading to false, loaded to false 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;

exports[`ProductCollectionReducer LOAD_SUCCESS  should set loading to false, loaded to true and set product ids [1, 2] 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [
    1,
    2,
  ],
  "loaded": true,
  "loading": false,
  "totalItems": 2,
}
`;

exports[`ProductCollectionReducer Selectors getAdded should retrieve added false 1`] = `false`;

exports[`ProductCollectionReducer Selectors getAdded should retrieve added true 1`] = `true`;

exports[`ProductCollectionReducer Selectors getAdding should retrieve adding false 1`] = `false`;

exports[`ProductCollectionReducer Selectors getAdding should retrieve adding true 1`] = `true`;

exports[`ProductCollectionReducer Selectors getConfig should retrieve config with page to 2 and itemsPerPage to 50 1`] = `
Object {
  "itemsPerPage": 50,
  "page": 2,
}
`;

exports[`ProductCollectionReducer Selectors getDeleted should retrieve deleted false 1`] = `false`;

exports[`ProductCollectionReducer Selectors getDeleted should retrieve deleted true 1`] = `true`;

exports[`ProductCollectionReducer Selectors getDeleting should retrieve removing false 1`] = `false`;

exports[`ProductCollectionReducer Selectors getDeleting should retrieve removing true 1`] = `true`;

exports[`ProductCollectionReducer Selectors getIds should retrieve ids [1,2,3] 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`ProductCollectionReducer Selectors getLoaded should retrieve loaded false 1`] = `false`;

exports[`ProductCollectionReducer Selectors getLoaded should retrieve loaded true 1`] = `true`;

exports[`ProductCollectionReducer Selectors getLoading should retrieve loading false 1`] = `false`;

exports[`ProductCollectionReducer Selectors getLoading should retrieve loading true 1`] = `true`;

exports[`ProductCollectionReducer Selectors getTotalItems should retrieve totalItems (100) 1`] = `100`;

exports[`ProductCollectionReducer undefined action should return the default state 1`] = `
Object {
  "added": false,
  "adding": false,
  "config": Object {
    "itemsPerPage": 10,
    "page": 1,
  },
  "deleted": false,
  "deleting": false,
  "ids": Array [],
  "loaded": false,
  "loading": false,
  "totalItems": 0,
}
`;
